File: backend/app/schemas/alert.py
from pydantic import BaseModel, field_validator, model_validator
from typing import Optional, Literal
from datetime import datetime
import re


class AlertCreate(BaseModel):
    ticker: str
    alert_type: Literal["price_above", "price_below", "earnings_date", "re_analyze", "custom"]
    target_price: Optional[float] = None
    cron_expression: Optional[str] = None
    event_description: Optional[str] = None

    @field_validator("ticker")
    @classmethod
    def validate_ticker(cls, v: str) -> str:
        v = v.strip().upper()
        if not re.match(r"^[A-Z]{1,5}(\.[A-Z]{1,2})?$", v):
            raise ValueError(f"Invalid ticker: {v}")
        return v

    @model_validator(mode="after")
    def validate_fields_by_type(self) -> "AlertCreate":
        if self.alert_type in ("price_above", "price_below") and self.target_price is None:
            raise ValueError("target_price is required for price alerts")
        if self.alert_type == "re_analyze" and not self.cron_expression:
            # Default: weekly on Monday at 8am
            self.cron_expression = "0 8 * * 1"
        if self.alert_type == "custom" and not self.event_description:
            raise ValueError("event_description is required for custom alerts")
        return self


class AlertUpdate(BaseModel):
    is_active: Optional[bool] = None
    target_price: Optional[float] = None
    cron_expression: Optional[str] = None
    event_description: Optional[str] = None


class AlertResponse(BaseModel):
    id: str
    ticker: str
    alert_type: str
    target_price: Optional[float] = None
    cron_expression: Optional[str] = None
    event_description: Optional[str] = None
    is_active: bool
    created_at: datetime
    last_triggered_at: Optional[datetime] = None
    trigger_count: int = 0

    class Config:
        from_attributes = True


class AlertTriggerResponse(BaseModel):
    id: str
    alert_id: str
    triggered_at: datetime
    trigger_value: str
    action_taken: str

    class Config:
        from_attributes = True

File: frontend/src/types/alert.ts
export type AlertType = 'price_above' | 'price_below' | 'earnings_date' | 're_analyze' | 'custom'

export interface Alert {
  id: string
  ticker: string
  alert_type: AlertType
  target_price?: number
  cron_expression?: string
  event_description?: string
  is_active: boolean
  created_at: string
  last_triggered_at?: string
  trigger_count: number
}

export interface AlertCreate {
  ticker: string
  alert_type: AlertType
  target_price?: number
  cron_expression?: string
  event_description?: string
}

export interface AlertTrigger {
  id: string
  alert_id: string
  triggered_at: string
  trigger_value: string
  action_taken: string
}

export const ALERT_TYPE_LABELS: Record<AlertType, string> = {
  price_above: '价格突破上限 Price Above',
  price_below: '价格跌破下限 Price Below',
  earnings_date: '财报日提醒 Earnings Date',
  re_analyze: '定期重新分析 Re-analyze',
  custom: '自定义事件 Custom',
}

export const FREQUENCY_OPTIONS = [
  { value: '0 8 * * 1', label: '每周一 Weekly (Monday)' },
  { value: '0 8 * * *', label: '每日 Daily' },
  { value: '0 8 1 * *', label: '每月1日 Monthly' },
]
